File: src/com/medium/SortCharactersByFrequency_451.java
/**
 * 
 */
package com.medium;

import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.PriorityQueue;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class SortCharactersByFrequency_451 {
    public static void main(String[] args) {
        String s = "BbAabb";
        frequencySort(s);
    }

    // Bucket Sort
    // Time complexity: O(n)
    // Space complexity: O(n)
    // Runtime: 14 ms, faster than 91.77%
    // Memory Usage: 38.7 MB, less than 97.17%
    public static String frequencySort(String s) {
        Map<Character, Integer> mMap = new HashMap<Character, Integer>();
        // for(int i = 0; i < mAnswer.length; i++) {
        // mMap.put(mAnswer[i], mMap.getOrDefault(mMap.get(mAnswer[i]), 0)+1);
        // }
        for (char c : s.toCharArray()) {
            mMap.put(c, mMap.getOrDefault(c, 0) + 1);
        }

        List<Character>[] mFrequencyBucket = new ArrayList[s.length() + 1];
        for (char c : mMap.keySet()) {
            int mCount = mMap.get(c);
            if (mFrequencyBucket[mCount] == null) {
                mFrequencyBucket[mCount] = new ArrayList<Character>();
            }
            mFrequencyBucket[mCount].add(c);
        }
        StringBuilder mStrAnswer = new StringBuilder();
        for (int i = mFrequencyBucket.length - 1; i >= 0; i--) {
            if (mFrequencyBucket[i] != null) {
                for (char c : mFrequencyBucket[i]) { // list all of char in
                                                     // bucket
                    for (int j = 0; j < i; j++) { // add char according bucket
                                                  // number
                        mStrAnswer.append(c);
                    }
                }
            }

        }
        // System.out.println(mAnswer);

        return mStrAnswer.toString();
    }

    // Bucket Sort 2
    // Time complexity: O(n)
    // Space complexity: O(n)
    // Runtime: 17 ms, faster than 73.52%
    // Memory Usage: 39 MB, less than 85.19%
    public String frequencySort2(String s) {
        HashMap<Character, Integer> mMap = new HashMap<>();
        List<Character>[] mBucket = new List[s.length() + 1];
        StringBuilder mResult = new StringBuilder();
        for (char c : s.toCharArray()) {
            mMap.put(c, mMap.getOrDefault(c, 0) + 1);
        }
        for (char c : mMap.keySet()) {
            int mFreq = mMap.get(c);
            if (mBucket[mFreq] == null) {
                mBucket[mFreq] = new ArrayList<>();
            }
            mBucket[mFreq].add(c);
        }
        for (int i = mBucket.length - 1; i >= 0; i--) {
            if (mBucket[i] != null) {
                for (char c : mBucket[i]) {
                    for (int j = 0; j < i; j++) {
                        mResult.append(c);
                    }
                }
            }
        }

        return mResult.toString();
    }

    // Use Priority Queue
    // Time complexity: O(nlogm) = O(n)
    // Space complexity: O(n)
    // m: 26 characters
    // Runtime: 44 ms, faster than 41.91%
    // Memory Usage: 37.9 MB, less than 96.30%
    public String frequencySort3(String s) {
        HashMap<Character, Integer> mMap = new HashMap<>();
        for (char c : s.toCharArray()) {
            mMap.put(c, mMap.getOrDefault(c, 0) + 1);
        }
        PriorityQueue<Character> mQueue = new PriorityQueue<>((x, y) -> mMap.get(y) - mMap.get(x));
        for (char c : mMap.keySet()) {
            mQueue.add(c);
        }
        StringBuilder mResult = new StringBuilder();
        while (!mQueue.isEmpty()) {
            char c = mQueue.poll();
            int mFreq = mMap.get(c);
            for (int i = 0; i < mFreq; i++) {
                mResult.append(c);
            }
        }
        return mResult.toString();
    }
}

File: Golang_src/medium/sortCharactersByFrequency_451.go
package medium

import "sort"

// Use hashmap and sort:
// Time Complexity: O(nlogn)
// Space Complexity:O(n)
// Runtime: 8 ms, faster than 57.81%
// Memory Usage: 6 MB, less than 26.56%
func frequencySort(s string) string {
	hashmap := map[rune]int{}
	chs := make([]rune, 0)

	for _, ch := range s {
		if value, exist := hashmap[ch]; exist {
			hashmap[ch] = value + 1
		} else {
			hashmap[ch] = 1
			chs = append(chs, ch)
		}
	}

	sort.Slice(chs, func(i, j int) bool {
		return hashmap[chs[i]] > hashmap[chs[j]]
	})

	result := []rune{}

	for _, v := range chs {
		freq := hashmap[v]
		for j := 0; j < freq; j++ {
			result = append(result, v)
		}
	}

	return string(result)
}

// Use Multiset and Bucket Sort:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 3 ms, faster than 94.78%
// Memory Usage: 5.8 MB, less than 41.74%
func frequencySor2(s string) string {
	hashmap := map[byte]int{}

	for i := range s {
		hashmap[s[i]]++
	}

	hashmap2 := map[int][]byte{}

	for k, v := range hashmap {
		if val, ok := hashmap2[v]; ok {
			val = append(val, k)
			hashmap2[v] = val
		} else {
			hashmap2[v] = []byte{k}
		}
	}

	result := []byte{}
	cur := len(s)
	for len(result) < len(s) {
		if val, ok := hashmap2[cur]; ok {
			for _, ch := range val {
				for j := 0; j < cur; j++ {
					result = append(result, ch)
				}
			}
		}
		cur--
	}

	return string(result)
}
